File: EPI/8.4/swift/8.4_test.swift
import Darwin

func nodes(_ nexts: [Int?]) -> [Node] {
    let nodes = (0..<nexts.count).map { _ in Node() }
    for (i, next) in nexts.enumerated() where next != nil {
        nodes[i].next = nodes[next!]
    }
    return nodes
}

let tests: [(nexts: [Int?], cycleStartIndex: Int?)] = [
    ([nil], nil),
    ([1, 2, nil], nil),
    ([0], 0),
    ([1, 0], 0),
    ([1, 2, 1], 1),
    ([1, 2, 3, 4, 5, 6, 0], 0),
    ([1, 2, 3, 4, 5, 6, 2], 2),
    ([1, 2, 3, 4, 5, 6, 4], 4),
    ([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 6], 6),
]

for test in tests {
    let n = nodes(test.nexts)
    let actualCycleStart = n[0].cycleStart()
    let actual = actualCycleStart == nil ? nil : n.index { $0 === actualCycleStart! }

    guard actual == nil && test.cycleStartIndex == nil ||
        actual != nil && test.cycleStartIndex != nil && actual! == test.cycleStartIndex! else {
        print("For nexts \(test.nexts), " +
            "expected cycleStartIndex to be \(test.cycleStartIndex as Optional), " +
            "but was \(actual as Optional)")
        exit(1)
    }
}

File: EPI/8.4/typescript/8.4_test.ts
import {Node, getFirstNodeInCycle} from './8.4';

declare function require(name: String): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

type Test = {list: Node, firstNodeInCycle: Node | undefined};

const tests: Test[] = [];

{
    // 0 -> 0
    const list: Node = {};
    list.next = list;
    tests.push({list, firstNodeInCycle: list});
}

{
    // 0 -> 1 -> 0
    const list: Node = {next: {}};
    list.next!.next = list;
    tests.push({list, firstNodeInCycle: list});
}

{
    // 0 -> 1 -> 2 -> 0
    const list: Node = {next: { next: {}}};
    list.next!.next!.next = list;
    tests.push({list, firstNodeInCycle: list});
}

{
    // 0 -> 1 -> 2 -> 1
    const list: Node = {next: { next: {}}};
    list.next!.next!.next = list.next;
    tests.push({list, firstNodeInCycle: list});
}

{
    // 0 -> undefined
    tests.push({list: {}, firstNodeInCycle: undefined});
}

{
    // 0 -> 1 -> undefined
    tests.push({list: {next: {}}, firstNodeInCycle: undefined});
}

for (const test of tests) {
    const actual = getFirstNodeInCycle(test.list);
    assert.deepStrictEqual(actual, test.firstNodeInCycle, `For test list ${inspect(test.list)}, ` +
        `expected first node in cycle to be ${inspect(test.firstNodeInCycle)}, ` +
        `but was ${inspect(actual)}`);
}
